File: ParkingRegisterAPI/Controllers/ParkingRegisterController.cs
using Microsoft.AspNetCore.Mvc;
using ParkingRegisterAPI.Models;
using ParkingRegisterAPI.ParkingRegistry;
using ParkingRegisterAPI.Responses;

namespace ParkingRegisterAPI.Controllers
{
    [ApiController]
    [Route("[controller]")]
    public class ParkingRegisterController : ControllerBase
    {
        private const int __MAXREQUESTS = 10;
        private const int __TIMEWINDOW = 10;

        private readonly string __slotNotFoundMsg = "no slot exists";
        private readonly string __carNotFoundMsg = "car not found";
        private readonly string __carparkFullMsg = "car park is full";
        private readonly string __invalidCarNumMsg = "no valid car number given";
        private readonly string __carAlreadyInSlotMsg = "car is already parked in a slot";
        private readonly string __slotIsEmptyPlaceholderMsg = "slot {0} has no car";
        private readonly string __slotHasNoCarPlaceholderMsg = "slot {0} does not have car with number {1}";
        private readonly string __slotNumInvalidMsg = "slot number is invalid";
        private readonly string __carSlotNumInvalidMsg = "car and slot numbers are invalid";

        private ISlotRegistry __slotRegistry;
        private ICarRegistry __carRegistry;

        public ParkingRegisterController(ISlotRegistry slotRegistry, ICarRegistry carRegistry)
        {
            __slotRegistry = slotRegistry;
            __carRegistry = carRegistry;
        }

        [HttpPost("ParkCar")]
        [LimitRequests(MaxRequests = __MAXREQUESTS, TimeWindow = __TIMEWINDOW)]
        public ParkCarResponse ParkCar(string carNumber)
        {
            if (string.IsNullOrWhiteSpace(carNumber)) return new ParkCarResponse { ErrorMessage = __invalidCarNumMsg };
            if (__carRegistry.GetCar(carNumber) != null) return new ParkCarResponse { ErrorMessage = __carAlreadyInSlotMsg };
            // find lot and park if exists
            string slotNumber = __slotRegistry.FindEmptySlotNumber();
            if (slotNumber == null) return new ParkCarResponse { ErrorMessage = __carparkFullMsg };
            // register car and it's lot
            Car newCar = new Car
            {
                Number = carNumber,
                Slot = __slotRegistry.GetSlot(slotNumber)
            };
            __slotRegistry.ParkCar(slotNumber, newCar);
            __carRegistry.RegisterCar(newCar);
            return new ParkCarResponse { SlotNumber = slotNumber };
        }

        [HttpDelete("UnparkCar")]
        [LimitRequests(MaxRequests = __MAXREQUESTS, TimeWindow = __TIMEWINDOW)]
        public UnparkCarResponse UnparkCar(string slotNumber)
        {
            if (string.IsNullOrWhiteSpace(slotNumber)) return new UnparkCarResponse { ErrorMessage = __slotNumInvalidMsg };
            Slot slot = __slotRegistry.GetSlot(slotNumber);
            if (slot == null) return new UnparkCarResponse { ErrorMessage = __slotNotFoundMsg };
            if (slot.IsEmpty()) return new UnparkCarResponse { ErrorMessage = String.Format(__slotIsEmptyPlaceholderMsg, slotNumber) };
            string carNumber = slot.Car.Number;
            __slotRegistry.UnparkCar(slotNumber);
            __carRegistry.DeregisterCar(carNumber);
            return new UnparkCarResponse { Success = true };
        }

        [HttpGet("CarSlotInfo")]
        [LimitRequests(MaxRequests = __MAXREQUESTS, TimeWindow = __TIMEWINDOW)]
        public CarSlotInfoResponse CarSlotInfo(string? slotNumber, string? carNumber)
        {
            if (string.IsNullOrWhiteSpace(slotNumber))
            {
                if (string.IsNullOrWhiteSpace(carNumber)) return new CarSlotInfoResponse { ErrorMessage = __carSlotNumInvalidMsg };
                Car car = __carRegistry.GetCar(carNumber);
                if (car == null) return new CarSlotInfoResponse { ErrorMessage = __carNotFoundMsg };
                return new CarSlotInfoResponse { SlotNumber = car.Slot.Number, CarNumber = carNumber };
            }

            Slot slot = __slotRegistry.GetSlot(slotNumber);
            if (slot == null) return new CarSlotInfoResponse { ErrorMessage = __slotNotFoundMsg };
            string carNumberFromSlot = slot.IsEmpty() ? null : slot.Car.Number;
            if (!string.IsNullOrEmpty(carNumber) && carNumberFromSlot != carNumber) 
                return new CarSlotInfoResponse { ErrorMessage = String.Format(__slotHasNoCarPlaceholderMsg, slot.Number, carNumber) };
            return new CarSlotInfoResponse { SlotNumber = slotNumber, CarNumber = carNumberFromSlot };
        }
    }
}

File: ParkingRegisterAPIFS/Controllers/ParkingRegisterController.fs
namespace ParkingRegisterAPIFS.Controllers

open System
open System.Collections.Generic
open System.Linq
open System.Threading.Tasks
open Microsoft.AspNetCore.Mvc
open Microsoft.Extensions.Logging
open ParkingRegisterAPIFS
open ParkingRegisterAPIFS.Models.Car
open ParkingRegisterAPIFS.Models.Slot
open ParkingRegisterAPIFS.Models.CarInSlot
open ParkingRegisterAPIFS.ParkingRegistry.SlotRegistry
open ParkingRegisterAPIFS.ParkingRegistry.CarRegistry
open ParkingRegisterAPIFS.Responses.ParkCarResponse
open ParkingRegisterAPIFS.Responses.UnparkCarResponse
open ParkingRegisterAPIFS.Responses.CarSlotInfoResponse

[<ApiController>]
[<Route("[controller]")>]
type ParkingRegisterController (slotRegistry: ISlotRegistry, carRegistry: ICarRegistry) =
    inherit ControllerBase()

    let __slotNotFoundMsg: string = "no slot exists";
    let __carNotFoundMsg: string = "car not found";
    let __carparkFullMsg: string = "car park is full";
    let __invalidCarNumMsg: string = "no valid car number given";
    let __carAlreadyInSlotMsg: string = "car is already parked in a slot";
    let __slotIsEmptyPlaceholderMsg: string = "slot {0} has no car";
    let __slotHasNoCarPlaceholderMsg: string = "slot {0} does not have car with number {1}";
    let __slotNumInvalidMsg: string = "slot number is invalid";
    let __carSlotNumInvalidMsg: string = "car and slot numbers are invalid";
    
    let __slotRegistry = slotRegistry;
    let __carRegistry = carRegistry;
    
    let getInfoByCarNumber(carNumber: string): CarSlotInfoResponse =
        let car: Nullable<Car> = __carRegistry.getCar(carNumber)
        if String.IsNullOrWhiteSpace carNumber then
            CarSlotInfoResponse("", "", __carSlotNumInvalidMsg)
        elif not car.HasValue then
            CarSlotInfoResponse("", "", __carNotFoundMsg)
        else
            CarSlotInfoResponse(car.Value.CarInSlot.Value.SlotNumber, carNumber, "")
    
    let getInfoBySlotNumberCheckCarNumber(slotNumber: string, carNumber: string): CarSlotInfoResponse =
        let slot: Nullable<Slot> = __slotRegistry.getSlot(slotNumber)
        if not slot.HasValue then
            let carNumberFromSlot: string =
                if slot.HasValue && slot.Value.CarInSlot.HasValue then slot.Value.CarInSlot.Value.CarNumber else ""
            if not (String.IsNullOrEmpty(carNumber)) && carNumberFromSlot <> carNumber then
                CarSlotInfoResponse("", "", String.Format(__slotHasNoCarPlaceholderMsg, slot.Value.Number, carNumber))
            else
                CarSlotInfoResponse(slotNumber, carNumberFromSlot, "")
        else
            CarSlotInfoResponse("", "", __slotNotFoundMsg)

    [<HttpPost("ParkCar")>]
    member _.parkCar(carNumber: string) =
        let slotNumber: string = __slotRegistry.findEmptySlotNumber();
        if String.IsNullOrWhiteSpace(carNumber) then
            ParkCarResponse("", __invalidCarNumMsg)
        elif __carRegistry.getCar(carNumber).HasValue then
            ParkCarResponse("", __carAlreadyInSlotMsg)
        elif String.IsNullOrEmpty(slotNumber) then
            ParkCarResponse("", __carparkFullMsg)
        else
            let newCar = Car(carNumber, CarInSlot(carNumber, slotNumber))
            __slotRegistry.parkCar(slotNumber, newCar)
            __carRegistry.registerCar(newCar)
            ParkCarResponse(slotNumber, "")

    [<HttpDelete("UnparkCar")>]
    member _.unparkCar(slotNumber: string) =
        let slot: Nullable<Slot> = __slotRegistry.getSlot(slotNumber)
        if String.IsNullOrWhiteSpace slotNumber then
            UnparkCarResponse(false, __slotNumInvalidMsg)
        elif not slot.HasValue then
            UnparkCarResponse(false, __slotNotFoundMsg)
        elif not slot.Value.CarInSlot.HasValue then
            UnparkCarResponse(false, String.Format(__slotIsEmptyPlaceholderMsg, slotNumber))
        else
            let carNumber = slot.Value.CarInSlot.Value.CarNumber
            __slotRegistry.unparkCar(slotNumber)
            __carRegistry.deregisterCar(carNumber)
            UnparkCarResponse(true, "")

    [<HttpGet("CarSlotInfo")>]
    member _.carSlotInfo(slotNumber: string, carNumber: string) =
        let slot: Nullable<Slot> = __slotRegistry.getSlot(slotNumber)
        if String.IsNullOrWhiteSpace slotNumber then
            getInfoByCarNumber(carNumber)
        else
            getInfoBySlotNumberCheckCarNumber(slotNumber, carNumber)
